src/action.ml

```diff
@@ -355,7 +356,12 @@ module Unexpanded = struct
       SW.partial_expand template ~f:(fun loc var ->
         match f loc var with
         | None   -> None
-        | Some e -> Some (VE.to_string ~dir e))
+        | Some e ->
+           try Some (VE.to_string ~dir e)
+           with Failure _ ->
+             let msg = sprintf "Split variable %S in string does not \
+                                make sense" var in
```

Comment:
Given the message, I'd match an explicit exception just to guard against some future additional failure mode.

I know there's another message which refers to things not making sense, but I think something more descriptive of what doesn't make sense would be better, e.g. `"${!%S} is a list and so cannot be used in a string (did you mean just ${%S}"`

It's also tempting just to say that `"${!@}" = "${@}"`

Reply:
> It's also tempting just to say that "${!@}" = "${@}"

Indeed!  (My take on this is that splitable variables should come be default as such — so that, say, `(run ${CC}` always does what is expected — and are concatenated with spaces if used in a string — e.g. `(run "${CC}"` to have the old behavior) .